### PR TITLE
Extra validation

### PR DESCRIPTION
**Pull request by Horneth**
Fixes https://github.com/broadinstitute/cromwell/issues/2394

### CODE REVIEW

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 1]
```diff
@@ -525,6 +525,21 @@ object WdlNamespace {
             Option(new SyntaxError(wdlSyntaxErrorFormatter.undefinedMemberAccess(memberAccessAst)))
         }
       }
+
+      val identifiers = ast.getAttribute("value").findSingleIdentifiers()
+      // Validates single identifiers. They can be declarations but not calls or workflows
+      val identifierValidation = identifiers flatMap { singleIdentifier =>
+          val resolvedScope = call.resolveVariable(singleIdentifier.sourceString)
```
Comment (mcovarr) [inline src/main/scala/wdl4s/WdlNamespace.scala]: indent is goofed up here

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 2]
```diff
@@ -525,6 +525,21 @@ object WdlNamespace {
...
+        resolvedScope match {
+          case Some(_: Call) =>
+            Option(new SyntaxError(wdlSyntaxErrorFormatter.illegalIdentifier(singleIdentifier)))
+          case None =>
```
Comment (mcovarr) [inline src/main/scala/wdl4s/WdlNamespace.scala]: this case corresponds to a workflow?

File: src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala [hunk 3]
```diff
@@ -185,6 +185,26 @@ case class WdlSyntaxErrorFormatter(terminalMap: Map[Terminal, WdlSource]) extend
...
+     |Cannot resolve variable ${terminal.getSourceString}
+     """.stripMargin
+  }
+
+  def illegalIdentifier(terminal: Terminal): String = {
+    s"""ERROR: Expression will not evaluate (line ${terminal.getLine}, col ${terminal.getColumn}):
+     |
+     |${pointToSource(terminal)}
+     |
+     |Calls and workflows cannot be used directly as inputs. Did you mean to use one of their output instead ?
```
Comment (mcovarr) [inline src/main/scala/wdl4s/WdlSyntaxErrorFormatter.scala]: output => outputs

**Comment by mcovarr** [comment]
:+1:

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 2 above)
Comment (Horneth) [inline src/main/scala/wdl4s/WdlNamespace.scala]: No it's if we can't find any scope that corresponds to this identifier.
The first cases encompasses both TaskCall and WorkflowCall

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 4]
```diff
@@ -945,6 +947,44 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
...
+        |}
+      """.stripMargin
+
+    val errors =
+      """ERROR: Expression will not evaluate (line 22, col 34):
+        |
+        |    call bye { input: instring = hello }
+        |                                 ^
+        |
+        |Calls and workflows cannot be used directly as inputs. Did you mean to use one of their outputs instead ?
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Extra space before `?`

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 5]
```diff
@@ -525,6 +525,21 @@ object WdlNamespace {
             Option(new SyntaxError(wdlSyntaxErrorFormatter.undefinedMemberAccess(memberAccessAst)))
         }
       }
+
+      val identifiers = ast.getAttribute("value").findSingleIdentifiers()
+      // Validates single identifiers. They can be declarations but not calls or workflows
+      val identifierValidation = identifiers flatMap { singleIdentifier =>
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: ToL: I sort of dislike `validation` in names that aren't cats validations

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 6]
```diff
@@ -945,6 +947,44 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
       """.stripMargin
   }
 
+  case object InvalidInputMapping extends ErrorWdl {
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Could we also have a test for the new `undefinedIdentified` formatter? Or even also include an undefined identifier in this one and check that we get both error messages?

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 5 above)
Comment (Horneth) [inline src/main/scala/wdl4s/WdlNamespace.scala]: suggestion ? 😄

Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: `identifierLinkErrors`

**Comment by Horneth** [comment]
I'm going to close this as I realized that although we accumulate all WDL validation errors (which is good) we only throw the first one of the list (which is a shame: https://github.com/broadinstitute/wdl4s/blob/develop/src/main/scala/wdl4s/WdlNamespace.scala#L410).. I'd like to change that but I won't have time to do it now so I'll re-open when I come back (if no one has done it in the meantime).

**Pull request closed without merging**